draft: Count and Bucket the Mempool

gregorycoppola opened the pull request:
Draft:  Based on #3337, solves #3188. 

Description:
* add a new `RelayerDirective` for `CountMempool`
* *followers* can set a flag `run_heavy_prometheus_analysis` to issue this directive, and periodically count their mempool and output to prometheus

Working:
* follower can issue the new `CountMempool` directive, and the count gets done
* follower outputs count results to `info!` log

Not yet working:
* use flag guarding
* actually output to prometheus

Commit:
stash

Commit:
added bucket_candidates

Commit:
it compiles!

Commit:
error[E0599]: no function or associated item named `get_nonce` found for struct `stacks::db::StacksChainState` in the current scope
   --> src/core/mempool.rs:821:47
    |
821 | ...                   StacksChainState::get_nonce(clarity_tx, &address.clone().into());
    |                                         ^^^^^^^^^ function or associated item not found in `stacks::db::StacksChainState`
    |
   ::: src/chainstate/stacks/db/mod.rs:109:1
    |
109 | pub struct StacksChainState {
    | --------------------------- function or associated item `get_nonce` not found for this

For more information about this error, try `rustc --explain E0599`.

Commit:
closer to working

Commit:
added `bucket_candidates`, which should work

Commit:
re-wrote loop so it compiles

Commit:
bucket_count_mempool makes sense

Commit:
stacks-inspect is compiling with `count-mempool` feature

Commit:
revert neon node

Commit:
fix comment

obycode commented on src/core/mempool.rs:
```diff
@@ -1129,6 +1290,79 @@ impl MemPoolDB {
+                }
+            }
+
+            // Bump nonces in the cache for the executed transaction
+            nonce_cache.update(
+                tx_info.origin_address,
+                expected_origin_nonce + 1,
+                self.conn(),
+            );
+            if tx_info.sponsor_address != tx_info.origin_address {
+                nonce_cache.update(
+                    tx_info.sponsor_address,
+                    expected_sponsor_nonce + 1,
+                    self.conn(),
+                );
+            }
```
I don't think you should bump the nonces in here. We should just be looking at transactions that are ready now.

Commit:
debugging why transactions get processed

Commit:
add back trace

Commit:
Merge branch 'develop' into draft/count-mempool-v2

Commit:
should be working now

Commit:
remove backtrace

Commit:
send the "count mempool signal"

Commit:
change the working dir

Commit:
Merge branch 'draft/count-mempool--integrate1' of github.com:gregorycoppola/stacks-blockchain into draft/count-mempool--integrate1

Commit:
more logs

Commit:
Merge branch 'draft/count-mempool--integrate1' of github.com:gregorycoppola/stacks-blockchain into draft/count-mempool--integrate1

Commit:
directive log

Commit:
fix the buid

Commit:
more logs

Commit:
Merge branch 'draft/count-mempool--integrate1' of github.com:gregorycoppola/stacks-blockchain into draft/count-mempool--integrate1

Commit:
send relayer signal earlier for a follower

Commit:
add in bucket_count_mempool into CountMempool directive

    Checking stacks-node v0.1.0 (/Users/greg/main1/testnet/stacks-node)
error[E0425]: cannot find value `parent_header` in this scope
    --> testnet/stacks-node/src/neon_node.rs:1238:26
     |
1238 |                         &parent_header,
     |                          ^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `chain_tip` in this scope
    --> testnet/stacks-node/src/neon_node.rs:1239:25
     |
1239 |                         chain_tip.total_burn,
     |                         ^^^^^^^^^ help: a local variable with a similar name exists: `chain_id`

Commit:
just missing parent header

error[E0425]: cannot find value `parent_header` in this scope
    --> testnet/stacks-node/src/neon_node.rs:1238:26
     |
1238 |                         &parent_header,
     |                          ^^^^^^^^^^^^^ not found in this scope

Commit:
compiles.. does it work?

gregorycoppola commented:
We definitely want to check this in. But, I'll close until progress is made.

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
